Add article about ClickHouse indexing and data processing

anelook opened the pull request:
This is information that I cut from the blog post, as it dives a bit deeper into how ClickHouse works.
I shaped it to focus on the primary index and data processing.

ftisiot commented on docs/products/clickhouse/concepts/indexing.rst:
```diff
@@ -0,0 +1,21 @@
+In this article we look at how ClickHouse works with data and how this differs from other database management systems.
+In particular we'll touch on such topics as **data indexing** and **a vector computation engine**.
+
+Understanding ClickHouse approach to data processing will help you learn why ClickHouse is perfect solution for some use cases and an awful choice for others.
```
should this be a tip?

anelook replied:
Changed

ftisiot commented on docs/products/clickhouse/concepts/indexing.rst:
```diff
@@ -0,0 +1,21 @@
+In particular we'll touch on such topics as **data indexing** and **a vector computation engine**.
+
+Understanding ClickHouse approach to data processing will help you learn why ClickHouse is perfect solution for some use cases and an awful choice for others.
+
```
I would put a subtitle here like `Columnar Database` and expand a bit.

You might compare it with the traditional databases storing a row at the time.

anelook replied:
We have a separate article on it. I can add more, but then I'm duplicating it... Added a clear reference to the separate article. Do you think this makes it more clear?

if not, I can indeed add some more sentences about columnar db

lornajane replied:
I think the link addresses what is needed here

ftisiot commented on docs/products/clickhouse/concepts/indexing.rst:
```diff
@@ -0,0 +1,21 @@
+
+ClickHouse is designed to process massive chunks of information for a single request. To make this possible it shifts the focus from reading individual lines into scanning massive blocks of data for a request. This happens by using what is called **a vector computation engine**, which reads data in blocks that normally consist of thousands of rows, selected for a small set of columns.
+
+To find a specific block, ClickHouse uses **a primary index**. However, unlike in other systems the primary index in ClickHouse has a couple of distinctive characteristics.
```
I would add a subtitle like `ClickHouse primary index` and put the ordering sentence before, since the primary index mechanism described before works only if the column is ordered beforehand

lornajane replied:
I am lost as well, so I'm resolving the comment and hoping for the best!

ftisiot commented on docs/products/clickhouse/concepts/indexing.rst:
```diff
@@ -0,0 +1,21 @@
+
+To find a specific block, ClickHouse uses **a primary index**. However, unlike in other systems the primary index in ClickHouse has a couple of distinctive characteristics.
+
+First of all, not every row is indexed. By contrast, indexing happens at every ten thousandth's row (to be precise, **8192nd** is the default suggested value). Such type of index is descriptively called **a sparse index**.
```
```suggestion
First of all, not every row is indexed. By contrast, indexing happens for one row every 10.000 (to be precise, **1/8192** is the default suggested value). Such type of index is descriptively called **a sparse index**.
```

lornajane replied:
This is better

ftisiot commented on docs/products/clickhouse/concepts/indexing.rst:
```diff
@@ -0,0 +1,21 @@
+
+First of all, not every row is indexed. By contrast, indexing happens at every ten thousandth's row (to be precise, **8192nd** is the default suggested value). Such type of index is descriptively called **a sparse index**.
+
+Using sparse indexing has a significant consequence on advantages and limitations of ClickHouse, mainly because such primary key does not ensure uniqueness for a single searched item. Seeing that we index every ten thousandth's item, in order to find a specific individual row we need to iterating over thousands of items. That's why this approach is inadequate when working with individual rows and ...
```
Can you state a bit more a good example of this? With the current example I probably get where I shouldn't use it only.

Maybe we could use: `Select * from TBL where column > 42`, the index is a good choice. 
`Select * from TBL where column = 42` probably the index is not a good choice since it still needs to scan the whole block

lornajane replied:
I _think_ I understand this now

ftisiot reviewed (changes requested):
Hi Olena, good piece and I def learnt something about ClickHouse. I requested few changes mainly about:
* Columnar database: probably explaining it against the usual row based databases makes the concept more digestible
* More examples on the indexing: to make clear when is a good fit and when not
* The order of sentences for the index: the primary index works only on top of a sorted column, so I believe we need to write the prereq first (sorted column) and then what you can do with it (primary index)

anelook commented:
Thx Francesco!! helpful comments!
I added some changes to improve content.

Regarding "the primary index works only on top of a sorted column" - this is a tricky point, where I'm  trying to avoid misleading the readers. So I wonder how to explain it to stay technically correct.  This is a territory where I'm not super confident that I'm getting all details correct. But I updated a bit the paragraph about it.

anelook commented:
Also, decided that I cannot ignore the secondary index and added some short overview on it as well.
This is by no mean deep analysis, I referenced official docs for that

lornajane reviewed (approved):
I learned something, thanks!
